_posts/2020-02-06-pwa-cache-me-if-you-can.markdown

```diff
@@ -0,0 +1,246 @@
+Check out Workbox [here.](https://developers.google.com/web/tools/workbox)
+
+```js
+importScripts('https://storage.googleapis.com/workbox-cdn/releases/4.3.1/workbox-sw.js');
+workbox.setConfig({ debug: false });
+workbox.precaching.precacheAndRoute([])
+
+workbox.routing.registerRoute(
+    /\.(?:png|jpg|jpeg|svg|gif)$/, //What image file types do you care about caching
+    new workbox.strategies.CacheFirst({
+        cacheName: 'image-cache',
+        plugins: [
+            new workbox.expiration.Plugin({
+                maxEntries: 20,
+                maxAgeSeconds: 7 * 24 * 60 * 60,
+            })
+        ],
+    })
+);
+```
+Whilst you are testing your caching strategy, you can omit the following line of code. I highly recommend that you add this in once you know your cache setup is working.
+
+```js
+workbox.setConfig({ debug: false }); 
+```
+
+### A2HS [Add 2 Home Screen]
+
+![A2HS](/img/blog/pwa/a2hs.png)
+
+You can prompt the user to install your PWA to their home screen from their desktop or mobile device, this allows the user to use your PWA as if it was a native app.
+
+```js
+let deferredPrompt;
+const addBtn = document.querySelector('.addButton') as HTMLElement;
+
+if (addBtn) {
+    addBtn.style.display = 'none';
+
+    window.addEventListener('beforeinstallprompt', (e) => {
+        // Prevent Chrome 67 and earlier from automatically showing the prompt
+        e.preventDefault();
+        // Stash the event so it can be triggered later.
+        deferredPrompt = e;
+        addBtn.style.display = 'block';
+
+        addBtn.addEventListener('click', (e) => {
+            // hide our user interface that shows our A2HS button
+            addBtn.style.display = 'none';
+            // Show the prompt
+            deferredPrompt.prompt();
+            deferredPrompt.userChoice.then((choiceResult) => {
+                if (choiceResult.outcome === 'accepted') {
+                    console.log('User accepted the A2HS prompt');
+                } else {
+                    console.log('User dismissed the A2HS prompt');
+                }
+                deferredPrompt = null;
+            });
+        });
+    });
+}
+```
+
+## Going Offline
+
+You can detect when your user is offline and display push notifications so the user is aware that they are no longer connected to the internet. 
+In my example below I was trying to take a photo when I lost internet connectivity, as my page was cached I was not navigated away from my page but notified that the image I had taken would be saved locally until I reconnected to the internet. Yes, I did put my thumb over the camera because you don't want to see my face.
+
+![offline](/img/blog/pwa/offline.png)
+
+
+### Wrapping up
+
+![audit](/img/blog/pwa/audit.png)
+
+If you think we can help you set up a PWA, feel free to [contact-us](https://www.mechanicalrock.io/lets-get-started).
+
+We will be at [ReactConf](https://reactconfau.com/), come say hi.
```

Comment:
need to release before ReactConf!

Reply:
I can remove this if it isn't ready

Reply:
I think we should aim to get it out before.. it seems ready?

Reply:
I have worked on Tim V's suggestions